File: IAIso-v5.0/core/iaiso-python/iaiso/middleware/gemini.py
"""Middleware for Google's Gemini (google-generativeai SDK) and Vertex AI.

Both SDKs expose `model.generate_content(...)` / `generate_content_async`.
Response objects have `usage_metadata.total_token_count` and
`candidates[].content.parts` that may include function-call parts.

This wrapper accounts for tokens from `usage_metadata` and counts
function-call parts as tool calls.

Install:  pip install iaiso[gemini]
"""

from __future__ import annotations

from typing import Any

from iaiso.core import BoundedExecution, ExecutionLocked, StepOutcome


class EscalationRaised(RuntimeError):
    pass


class GeminiBoundedModel:
    """Wraps a google.generativeai.GenerativeModel (or Vertex AI
    GenerativeModel) for pressure accounting.

    Example:
        >>> import google.generativeai as genai
        >>> from iaiso import BoundedExecution
        >>> from iaiso.middleware.gemini import GeminiBoundedModel
        >>>
        >>> model = genai.GenerativeModel("gemini-1.5-pro")
        >>> with BoundedExecution.start() as exec_:
        ...     wrapped = GeminiBoundedModel(model, exec_)
        ...     resp = wrapped.generate_content("hello")

    Works against both:
        - google.generativeai (public Gemini API)
        - vertexai.generative_models (Vertex AI)
    The response shape is the same across both.
    """

    def __init__(
        self,
        model: Any,
        execution: BoundedExecution,
        *,
        raise_on_escalation: bool = False,
    ) -> None:
        self._model = model
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation

    def generate_content(self, *args: Any, **kwargs: Any) -> Any:
        self._preflight()
        response = self._model.generate_content(*args, **kwargs)
        self._account(response)
        return response

    async def generate_content_async(self, *args: Any, **kwargs: Any) -> Any:
        self._preflight()
        response = await self._model.generate_content_async(*args, **kwargs)
        self._account(response)
        return response

    def start_chat(self, *args: Any, **kwargs: Any) -> "_BoundedChatSession":
        """Return a wrapper around a ChatSession."""
        inner = self._model.start_chat(*args, **kwargs)
        return _BoundedChatSession(inner, self._execution,
                                   self._raise_on_escalation)

    def __getattr__(self, name: str) -> Any:
        return getattr(self._model, name)

    def _preflight(self) -> None:
        pre = self._execution.check()
        if pre is StepOutcome.LOCKED:
            raise ExecutionLocked(
                f"execution {self._execution.engine.execution_id} is locked"
            )
        if pre is StepOutcome.ESCALATED and self._raise_on_escalation:
            raise EscalationRaised(
                f"execution escalated at pressure "
                f"{self._execution.engine.pressure:.3f}"
            )

    def _account(self, response: Any) -> None:
        tokens = 0
        meta = getattr(response, "usage_metadata", None)
        if meta is not None:
            tokens = int(getattr(meta, "total_token_count", 0))

        # Count function-call parts across all candidates
        tool_calls = 0
        candidates = getattr(response, "candidates", None) or []
        for cand in candidates:
            content = getattr(cand, "content", None)
            if content is None:
                continue
            parts = getattr(content, "parts", None) or []
            for part in parts:
                if getattr(part, "function_call", None) is not None:
                    tool_calls += 1

        model_name = str(getattr(self._model, "model_name", "gemini"))
        self._execution.record_step(
            tokens=tokens,
            tool_calls=tool_calls,
            tag=f"gemini.generate:{model_name}",
        )


class _BoundedChatSession:
    """Wraps a Gemini ChatSession so `send_message` is accounted."""

    def __init__(
        self,
        session: Any,
        execution: BoundedExecution,
        raise_on_escalation: bool,
    ) -> None:
        self._session = session
        self._execution = execution
        self._raise_on_escalation = raise_on_escalation
        self._parent = GeminiBoundedModel.__new__(GeminiBoundedModel)
        self._parent._model = None  # unused
        self._parent._execution = execution
        self._parent._raise_on_escalation = raise_on_escalation

    def send_message(self, *args: Any, **kwargs: Any) -> Any:
        self._parent._preflight()
        response = self._session.send_message(*args, **kwargs)
        self._parent._account(response)
        return response

    async def send_message_async(self, *args: Any, **kwargs: Any) -> Any:
        self._parent._preflight()
        response = await self._session.send_message_async(*args, **kwargs)
        self._parent._account(response)
        return response

    def __getattr__(self, name: str) -> Any:
        return getattr(self._session, name)

File: IAIso-v5.0/core/iaiso-node/src/middleware/gemini.ts
/**
 * Google Gemini / Vertex AI middleware.
 *
 * Wraps `@google/generative-ai` so calls to `model.generateContent()`
 * are accounted for in a `BoundedExecution`. Token counts come from
 * `response.usageMetadata`; tool/function calls are counted from
 * candidate parts of type `functionCall`.
 *
 * Install:
 *   npm install @google/generative-ai
 *
 * Example:
 *   import { GoogleGenerativeAI } from "@google/generative-ai";
 *   import { BoundedExecution, GeminiBoundedModel } from "@iaiso/core";
 *
 *   const genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY!);
 *   await BoundedExecution.run({ config, audit_sink }, async (exec) => {
 *     const raw = genAI.getGenerativeModel({ model: "gemini-1.5-pro" });
 *     const model = new GeminiBoundedModel(raw, exec);
 *     const result = await model.generateContent("hello");
 *   });
 */

import { BoundedExecution, ExecutionLocked } from "../core/execution.js";
import { StepOutcome } from "../core/types.js";
import { EscalationRaised } from "./anthropic.js";

interface GeminiResponse {
  response?: {
    usageMetadata?: {
      promptTokenCount?: number;
      candidatesTokenCount?: number;
      totalTokenCount?: number;
    };
    candidates?: Array<{
      content?: {
        parts?: Array<{ functionCall?: unknown; text?: string }>;
      };
    }>;
  };
  usageMetadata?: {
    promptTokenCount?: number;
    candidatesTokenCount?: number;
    totalTokenCount?: number;
  };
  candidates?: Array<{
    content?: {
      parts?: Array<{ functionCall?: unknown; text?: string }>;
    };
  }>;
}

interface GeminiGenerativeModelLike {
  generateContent: (
    request: string | object,
  ) => Promise<GeminiResponse> | GeminiResponse;
  model?: string;
}

export interface GeminiBoundedModelOptions {
  raiseOnEscalation?: boolean;
}

export class GeminiBoundedModel<
  T extends GeminiGenerativeModelLike = GeminiGenerativeModelLike,
> {
  readonly raw: T;
  private readonly _execution: BoundedExecution;
  private readonly _raiseOnEscalation: boolean;

  constructor(
    model: T,
    execution: BoundedExecution,
    opts: GeminiBoundedModelOptions = {},
  ) {
    this.raw = model;
    this._execution = execution;
    this._raiseOnEscalation = opts.raiseOnEscalation ?? false;
  }

  async generateContent(request: string | object): Promise<GeminiResponse> {
    const pre = this._execution.check();
    if (pre === StepOutcome.Locked) {
      throw new ExecutionLocked(
        `execution ${this._execution.engine.execution_id} is locked`,
      );
    }
    if (pre === StepOutcome.Escalated && this._raiseOnEscalation) {
      throw new EscalationRaised(
        `execution escalated at pressure ${this._execution.engine.pressure.toFixed(3)}`,
      );
    }
    const response = await this.raw.generateContent(request);
    this._account(response);
    return response;
  }

  private _account(resp: GeminiResponse): void {
    // Gemini responses sometimes nest under `.response`; accept both shapes
    const inner = resp.response ?? resp;
    const usage = inner.usageMetadata ?? {};
    const tokens =
      Number(usage.totalTokenCount ?? 0) ||
      Number(usage.promptTokenCount ?? 0) +
        Number(usage.candidatesTokenCount ?? 0);

    let toolCalls = 0;
    for (const cand of inner.candidates ?? []) {
      for (const part of cand.content?.parts ?? []) {
        if ((part as { functionCall?: unknown }).functionCall) toolCalls += 1;
      }
    }

    const model = this.raw.model ?? "gemini";
    this._execution.recordStep({
      tokens,
      tool_calls: toolCalls,
      tag: `gemini.generateContent:${model}`,
    });
  }
}
